prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm.c

```diff
@@ -642,52 +626,58 @@ void check_and_repost_receives(struct fi_ibv_rdm_ep *ep,
 	}
 }
 
-static inline void
-fi_ibv_rdm_tagged_got_recv_completion(struct fi_ibv_rdm_ep *ep,
-				      struct fi_ibv_rdm_tagged_conn *conn,
-				      int arrived_len, uint32_t imm_data)
-{
-	assert(arrived_len > 0);
-	conn->recv_preposted--;
-	check_and_repost_receives(ep, conn);
-	struct fi_ibv_rdm_buf *rbuf = fi_ibv_rdm_get_rbuf(conn, ep, imm_data);
-	fi_ibv_rdm_tagged_process_recv(ep, conn, arrived_len, imm_data, rbuf);
-}
-
 static inline int fi_ibv_rdm_tagged_poll_recv(struct fi_ibv_rdm_ep *ep)
 {
-	const int wc_count = ep->n_buffs;	// TODO: to set from upper level
+	const int wc_count = ep->n_buffs; /* TODO: to set from upper level */
 	struct ibv_wc wc[wc_count];
 	int i;
 	int ret;
 
 	do {
 		ret = ibv_poll_cq(ep->rcq, wc_count, wc);
 		for (i = 0; i < ret; ++i) {
-			if (wc[i].status == IBV_WC_SUCCESS
-			    && wc[i].opcode == IBV_WC_RECV_RDMA_WITH_IMM) {
+			if (wc[i].status == IBV_WC_SUCCESS) {
+				assert(wc[i].opcode == IBV_WC_RECV_RDMA_WITH_IMM);
 				FI_IBV_DBG_OPCODE(wc[i].opcode, "RECV");
 
 				struct fi_ibv_rdm_tagged_conn *conn =
-				    (struct fi_ibv_rdm_tagged_conn *)(uintptr_t)
-				    wc[i].wr_id;
+					(void *) wc[i].wr_id;
+
+				struct fi_ibv_rdm_buf *rbuf = fi_ibv_rdm_get_rbuf(conn, ep, conn->recv_processed);
+
+				FI_IBV_PREFETCH_ADDR(rbuf);
+
+				conn->recv_completions++;
+				if (conn->recv_completions & ep->n_buffs) {
+					conn->recv_completions = 0;
+				}
+
+				check_and_repost_receives(ep, conn);
+
+				if (fi_ibv_rdm_get_buffer_status((char*) rbuf) == BUF_STATUS_RECVED &&
+				    fi_ibv_rdm_buffer_check_pkt_len((char*) rbuf, wc[i].byte_len))
+				{
+					do {
+						fi_ibv_rdm_tagged_process_recv(ep, conn, fi_ibv_rdm_get_buff_service_data((char*) rbuf)->pkt_len, rbuf);
+						fi_ibv_rdm_set_buffer_status((char*) rbuf, BUF_STATUS_FREE);
```

Comment:
This is nested too deep.  Consider moving inner loops into separate functions for readability.  (The compiler will fold everything back in.)